controllers/onload_controller.go

```diff
@@ -56,15 +73,197 @@ func (r *OnloadReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctr
+		}
+
+		err := r.Update(ctx, onload)
+		if err != nil {
+			log.Error(err, "Failed to remove finalizer")
+			return false, ctrl.Result{}, err
+		}
+	}
+	return true, ctrl.Result{}, nil
+}
+
+func (r *OnloadReconciler) createAndAddModules(
+	ctx context.Context,
+	onload *onloadv1alpha1.Onload,
+) (bool, ctrl.Result, error) {
+	log := log.FromContext(ctx)
+	if !controllerutil.ContainsFinalizer(onload, onloadModuleFinalizer) {
+		log.Info("Adding module finalizer for onload")
+		if !controllerutil.AddFinalizer(onload, onloadModuleFinalizer) {
+			log.Info("Failed to add finalizer to custom resource")
+			return false, ctrl.Result{Requeue: true}, nil
+		}
+
+		err := r.Update(ctx, onload)
+		if err != nil {
+			log.Error(err,
+				"Failed to update custom resource with added finalizer")
+			return false, ctrl.Result{}, err
+		}
+	}
+
+	moduleNames := []string{
+		onload.Name + "-onload-module",
+		// onload.Name + "-sfc-module",
+	}
+
+	for _, moduleName := range moduleNames {
+		module := &kmm.Module{}
+		err := r.Get(
+			ctx,
+			types.NamespacedName{
+				Name:      moduleName,
+				Namespace: onload.Namespace,
+			},
+			module,
+		)
+		if err != nil && apierrors.IsNotFound(err) {
+			module, err := r.createModule(onload, "onload", moduleName)
+			if err != nil {
+				log.Error(err, "createModule failure")
+				return false, ctrl.Result{}, err
+			}
+
+			log.Info("Creating onload module")
+			err = r.Create(ctx, module)
+			if err != nil {
+				log.Error(err, "Failed to create new module")
+				return false, ctrl.Result{}, err
+			}
+
+			return false, ctrl.Result{RequeueAfter: time.Minute}, nil
+
+		} else if err != nil {
+			log.Error(err, "Failed to get onload-module")
+			return false, ctrl.Result{}, err
+		}
+	}
+
+	return true, ctrl.Result{}, nil
+}
+
+func (r *OnloadReconciler) createModule(
+	onload *onloadv1alpha1.Onload,
+	modprobeArg string,
+	moduleName string,
+) (*kmm.Module, error) {
+
+	kernelMappings := make(
+		[]kmm.KernelMapping,
+		len(onload.Spec.Onload.KernelMappings),
+	)
+
+	for _, kmapSpec := range onload.Spec.Onload.KernelMappings {
+		if kmapSpec.Build != nil {
+			return nil, errors.New(
+				"onload operator doesn't support non-nil build (yet)",
+			)
+		}
+
+		kmap := kmm.KernelMapping{
+			Regexp:         kmapSpec.Regexp,
+			ContainerImage: kmapSpec.KernelModuleImage,
+		}
+		kernelMappings = append(kernelMappings, kmap)
+	}
+
+	module := &kmm.Module{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      moduleName,
+			Namespace: onload.Namespace,
+		},
+		Spec: kmm.ModuleSpec{
+			ModuleLoader: kmm.ModuleLoaderSpec{
+				ServiceAccountName: onload.Spec.ServiceAccountName,
+				Container: kmm.ModuleLoaderContainerSpec{
+					Modprobe: kmm.ModprobeSpec{
+						ModuleName: modprobeArg,
+						Parameters: []string{"--first-time"},
+					},
+					KernelMappings:  kernelMappings,
+					ImagePullPolicy: onload.Spec.Onload.ImagePullPolicy,
+				},
+			},
+			Selector: map[string]string{"kubernetes.io/arch": "amd64"},
```

Comment:
```suggestion
			Selector: onload.Spec.Selector,
```

(Spotted because I'm filling in the DS spec and continuously checking with Module.)